include/mfmg/dealii/lanczos_deflatedlanczos.hpp

```diff
@@ -0,0 +1,89 @@
+/*************************************************************************
+ * Copyright (c) 2017-2019 by the mfmg authors                           *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef _LANCZOS_DEFLATEDLANCZOS_HPP_
+#define _LANCZOS_DEFLATEDLANCZOS_HPP_
+
+#include <vector>
+
+namespace mfmg::lanczos
+{
+
+//-----------------------------------------------------------------------------
+/// \brief Deflated Lanczos solver
+///
+///        The Lanczos solver is called multiple times.  After each Lanczos
+///        solve, the operator is modified to deflate out all previously
+///        computed (approximate) eigenvectors.  This is meant to deal with
+///        possible eigenvalue multiplicities.
+
+template<typename Op_t_>
+class DeflatedLanczos {
+
+  public:
+
+    // Typedefs
+
+    typedef Op_t_ Op_t;
+    typedef typename Op_t::Vector_t Vector_t;
+    typedef typename Op_t::Scalar_t Scalar_t;
+    typedef typename std::vector<Scalar_t> Scalars_t;
+    typedef typename std::vector<Vector_t*> Vectors_t;
+
+    // Ctor/dtor
+
+    DeflatedLanczos(Op_t& op, int num_evecs_per_cycle, int num_cycles,
+            int maxit, double tol, unsigned int percent_overshoot = 0,
+            unsigned int verbosity = 0);
+    ~DeflatedLanczos();
+
+    // Accessors
+
+    Scalar_t get_eval(int i) const;
+
+    Vector_t* get_evec(int i) const;
+
+    int num_evecs() const {return num_evecs_per_cycle_ * num_cycles_;}
+
+    // Operations
+
+    void solve();
+
+  private:
+
+    const Op_t& op_;
+    const int num_evecs_per_cycle_; // number of eigs to calc per lanc solve
+    const int num_cycles_;          // number of lanczos solves
+    const int maxit_;               // maximum number of lanc interations
+    const double tol_;              // convergence tolerance for eigenvalue
+    const unsigned int percent_overshoot_;
+                                    // allowed iteration count overshoot from
+                                    // less frequent stopping tests
```

Comment:
I am not sure what this means.

Reply:
@wdj

Reply:
A runtime-saving feature to avoid computing the eigenvalues of the tridiagonal to do the stopping test at every Lanczos step.  By computing every few steps you may save compute time but go a few more steps (no more than a fixed percentage) than needed.